flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImplTest.java

```diff
@@ -0,0 +1,357 @@
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+import org.apache.flink.runtime.shuffle.PartitionDescriptor;
+import org.apache.flink.runtime.shuffle.ShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.collection.IsEmptyCollection.empty;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Tests for the {@link JobAwareShuffleEnvironmentImpl}.
+ */
+public class JobAwareShuffleEnvironmentImplTest extends TestLogger {
+
+	private JobID jobId = new JobID();
+	private TestResultPartitionWriter externallyManagedWriter;
+	private TestResultPartitionWriter internallyManagedWriter;
+	private TestShuffleEnvironment wrappedEnvironment;
+	private JobAwareShuffleEnvironmentImpl<?> jobAwareEnvironment;
+	private Collection<ResultPartitionWriter> writers;
+
+	@Before
+	public void setup() {
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		externallyManagedWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+		internallyManagedWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+
+		wrappedEnvironment = new TestShuffleEnvironment(
+			Arrays.asList(externallyManagedWriter, internallyManagedWriter));
+
+		jobAwareEnvironment = new JobAwareShuffleEnvironmentImpl<>(wrappedEnvironment);
+
+		ResultPartitionDeploymentDescriptor deploymentDescriptor1 = createResultPartitionDeploymentDescriptor(
+			externallyManagedWriter.resultPartitionID, true);
+		ResultPartitionDeploymentDescriptor deploymentDescriptor2 = createResultPartitionDeploymentDescriptor(
+			internallyManagedWriter.resultPartitionID, false);
+
+		writers = createResultPartitionWriters(
+			jobAwareEnvironment,
+			jobId,
+			executionAttemptId,
+			deploymentDescriptor1,
+			deploymentDescriptor2
+		);
+	}
+
+	@Test
+	public void testInitialState() {
+		final JobID jobId = new JobID();
+		final TestResultPartitionWriter partitionWriter1 = new TestResultPartitionWriter(new ResultPartitionID());
+		final TestResultPartitionWriter partitionWriter2 = new TestResultPartitionWriter(new ResultPartitionID());
+
+		final TestShuffleEnvironment wrappedEnvironment = new TestShuffleEnvironment(
+			Arrays.asList(partitionWriter1, partitionWriter2));
+
+		final JobAwareShuffleEnvironmentImpl jobAwareEnvironment = new JobAwareShuffleEnvironmentImpl(wrappedEnvironment);
```

Comment:
we do not need to create all the above components, just use them in `setup`, so we could only retain `assertFalse(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));` for this method,